pilot/pkg/model/push_context.go

```diff
@@ -1708,12 +1708,26 @@ func (ps *PushContext) BestEffortInferServiceMTLSMode(service *Service, port *Po
 		return MTLSUnknown
 	}
 
-	// First , check mTLS settings from beta policy (i.e PeerAuthentication) at namespace / mesh level.
+	// 1. Check service instances' tls mode, mainly used for headless service.
+	if service.Resolution == Passthrough {
+		instances, _ := ps.InstancesByPort(service, port.Port, nil)
```

Comment:
we had concerns about the performance of this in the original PR and made some (major) changes to make this more performant.